src/llvm/mod.rs

```diff
@@ -179,69 +181,130 @@ pub unsafe fn optimize(
-
-    LLVMAddAnalysisPasses(tm, mpm);
-    LLVMAddAnalysisPasses(tm, fpm);
-
-    // even with -O0 and without LTO we still want to avoid linking in unused code from core etc
-    LLVMAddGlobalDCEPass(mpm);
-
-    let pmb = LLVMPassManagerBuilderCreate();
-
-    use OptLevel::*;
-    let (inline_threshold, opt) = match opt_level {
-        No | SizeMin => (0, 1), // Pretty much nothing compiles with -O0 s∫o make it an alias for -O1
-        Less => (25, 1),
-        Default => (225, 2),
-        Aggressive => (275, 3),
-        Size => (25, 0),
-    };
-    LLVMPassManagerBuilderSetOptLevel(pmb, opt);
-    LLVMPassManagerBuilderSetSizeLevel(
-        pmb,
-        match opt_level {
-            Size => 1,
-            SizeMin => 2,
-            _ => 0,
-        },
-    );
-    LLVMPassManagerBuilderUseInlinerWithThreshold(pmb, inline_threshold);
-
-    // populate the pass managers
-    LLVMPassManagerBuilderPopulateFunctionPassManager(pmb, fpm);
-    LLVMPassManagerBuilderPopulateModulePassManager(pmb, mpm);
-
     for sym in module.globals_iter() {
         internalize(sym, symbol_name(sym), export_symbols);
     }
     for sym in module.global_aliases_iter() {
         internalize(sym, symbol_name(sym), export_symbols);
     }
 
-    debug!("running function passes");
-    LLVMInitializeFunctionPassManager(fpm);
     for function in module.functions_iter() {
         let name = symbol_name(function);
         if !name.starts_with("llvm.") {
             if ignore_inline_never {
                 remove_attribute(function, "noinline");
             }
             internalize(function, name, export_symbols);
-            if LLVMIsDeclaration(function) == 0 {
-                LLVMRunFunctionPassManager(fpm, function);
-            }
         }
     }
-    LLVMFinalizeFunctionPassManager(fpm);
 
-    debug!("running module passes");
-    LLVMRunPassManager(mpm, module);
+    let options = LLVMCreatePassBuilderOptions();
+
+    // even with -O0 and without LTO we still want to avoid linking in unused code from core etc
+    let dce = CString::new("dce").unwrap();
+    debug!("running dce passe");
+    LLVMRunPasses(module, dce.as_ptr(), tm, options);
+
+    let opt = CString::new(match opt_level {
+        // Pretty much nothing compiles with -O0 so make it an alias for -O1
+        OptLevel::No | OptLevel::Less => "default<O1>",
+        OptLevel::Default => "default<O2>",
+        OptLevel::Aggressive => "default<O3>",
+        OptLevel::Size => "default<Os>",
+        OptLevel::SizeMin => "default<Oz>",
+    })
+    .unwrap();
+
+    debug!("running optimization pass");
+    LLVMRunPasses(module, opt.as_ptr(), tm, options);
+
+    LLVMDisposePassBuilderOptions(options);
 
     // Some debug info generated by rustc seems to trigger a segfault in the
     // BTF code in llvm, so strip it until that is fixed
     LLVMStripModuleDebugInfo(module);
+
+    // Collect up all the memory intrinsics. We're going to replace these with
```

Comment:
nit: this whole thing to rewrite the memory intrinsics feels like it could belong in its own function.

Reply:
It could. What does it get us? I'm not a fan of indirection for its own sake.

Reply:
Ack. I think it can improve readability in so far as a human can summarize what the function does. By having the logic to do this rewriting inline, it becomes more difficult to internalize the code structure from a single pass. I am not religious here.

Reply:
This code is no longer in this PR.